### PR TITLE
Fuzz v24

### PR DESCRIPTION
**Pull request by catenacyber**
Link to redmine ticket:
https://redmine.openinfosecfoundation.org/issues/2859

Describe changes:
- Adds eight fuzz targets
- Change compilation scheme so as to compile fuzz targets (option --enable-fuzztargets for configure)

Modifies #4614 with :
- use of `AFLFUZZ_PERSISTANT_MODE` enabled when configure has option `--enable-fuzztargets`

After this, we can 
- review and delete `#ifdef AFLFUZZ_PCAP_RUNMODE` and such
Who best can tell if we can remove these chunks of code ?
- Change .travis.yml  to make a build with `--enable-fuzztargets` configure option, then run the fuzz targets with sanitizers enabled
- Get the PR to oss-fuzz running : https://github.com/google/oss-fuzz/pull/2533 and fix the found bugs, add a status badge, try CIFuzz...

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -459,9 +459,49 @@
...
+
+    AC_ARG_ENABLE(fuzztargets,
+        AS_HELP_STRING([--enable-fuzztargets], [Enable fuzz targets]),[enable_fuzztargets=$enableval],[enable_fuzztargets=no])
+    AM_CONDITIONAL([BUILD_FUZZTARGETS], [test "x$enable_fuzztargets" = "xyes"])
+    AS_IF([test "x$enable_fuzztargets" = "xyes"], [
+        AC_DEFINE([AFLFUZZ_NO_RANDOM], [1], [Disable all use of random functions])
+        AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[]], [[while (__AFL_LOOP(1000))]])],
+                [AC_DEFINE([AFLFUZZ_PERSISTANT_MODE], [1], [Enable AFL PERSISTANT_MODE])],
+                [])
+        AC_PROG_CXX
```
Comment (jasonish) [inline configure.ac]: Defining this here results in some ugly output during autoconf:
```
../../lib/autoconf/c.m4:667: AC_LANG_COMPILER(C++) is expanded from...
../../lib/autoconf/lang.m4:329: AC_LANG_COMPILER_REQUIRE is expanded from...
../../lib/autoconf/general.m4:2678: AC_LINK_IFELSE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/autoconf/general.m4:2031: AC_CACHE_VAL is expanded from...
```
It looks like moving AC_PROG_CXX outside of the AS_IF fixes this. I believe the macro is expanded either way, and won't cause a failure even if no C++ compiler can be found.

Comment (catenacyber) [reply configure.ac]: But do we want to require a C++ compiler when it is only needed for fuzzing ?

File: configure.ac [hunk 2]
```diff
@@ -459,9 +459,49 @@
 
 # options
 
+
+    AC_ARG_ENABLE(fuzztargets,
+        AS_HELP_STRING([--enable-fuzztargets], [Enable fuzz targets]),[enable_fuzztargets=$enableval],[enable_fuzztargets=no])
+    AM_CONDITIONAL([BUILD_FUZZTARGETS], [test "x$enable_fuzztargets" = "xyes"])
+    AS_IF([test "x$enable_fuzztargets" = "xyes"], [
+        AC_DEFINE([AFLFUZZ_NO_RANDOM], [1], [Disable all use of random functions])
+        AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[]], [[while (__AFL_LOOP(1000))]])],
```
Comment (victorjulien) [inline configure.ac]: This has no effect as `config.h` is not included in `onefile.c`.

File: src/tests/fuzz/onefile.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+
```
Comment (victorjulien) [inline src/tests/fuzz/onefile.c]: Add
```c
#include "config.h"
```
For AFL persistent mode to work.

File: configure.ac (hunk 1 above)
Comment (catenacyber) [reply configure.ac]: You are right Jason, no failure even if no C++ compiler is found, thanks :-)

File: src/tests/fuzz/onefile.c (hunk 3 above)
Comment (catenacyber) [reply src/tests/fuzz/onefile.c]: ok thanks

File: configure.ac (hunk 2 above)
Comment (catenacyber) [reply configure.ac]: ok thanks

**Pull request closed without merging**